### PR TITLE
JS Publishing

### PR DESCRIPTION
**Pull request by dmdashenkov**
In this PR we configure NPM publishing for the `spine-js-client` package.

The package is published on each `master` merge, together with all the other artifacts.

Also in this PR:
 - Spine and Google proto definitions are extracted into a separate Node module and also published;
 - the Node modules are copied into an isolated directory for build and publish (this is required in order to prepare the package descriptor before passing to `npm`);
 - JS-related Gradle declarations are moved into a separate file - `js.gradle`; we may want to extract that into a standalone Gradle plugin someday;
 - JS source code now uses 4-space margin instead of 2-space;
 - the library now supports CORS requests.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: client-js/src/client/backend-client.js [hunk 1]
```diff
@@ -28,88 +28,83 @@
...
-   * @param actorRequestFactory the {@link ActorRequestFactory} to instantiate
-   *                            the actor requests with
-   */
-  constructor(httpClient, firebaseClient, actorRequestFactory) {
-    this._httpClient = httpClient;
-    this._firebase = firebaseClient;
-    this._actorRequestFactory = actorRequestFactory;
-  }
+    /**
+     * Creates a new BackendClient.
```
Comment (armiol) [inline client-js/src/client/backend-client.js]:
```
`BackendClient`
```

File: scripts/js.gradle [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
```
Comment (armiol) [inline scripts/js.gradle:20]: To me the name of this file is way too generic. 

Can you please add some general description of the purpose to this file, and judging on that, think of a better name for the file?

In particular, I'd like to understand what will happen after I use this script in a module, and why would I want to do that.

**Review by armiol — COMMENTED** [review]
@dmdashenkov please see my comment.

The rest LGTM.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

File: scripts/js.gradle [hunk 3]
```diff
@@ -18,6 +18,19 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+/**
+ * This build script contains a Gradle plugin for JavaScript modules.
+ *
+ * <p>The plugin configures NPM tools execution and connects the JS assembly and test routines
+ * with the Gradle execution process.
+ *
+ * <p>The plugin defines a number of tasks, which launch NPM scripts, compile Protobuf into
```
Comment (armiol) [inline scripts/js.gradle]: As discussed, please have an exact list of what to expect.

**Review by armiol — APPROVED** [review]
@dmdashenkov LGTM.

**Pull request merged**